astropy/utils/data_info.py

```diff
@@ -338,25 +340,21 @@ def _parent(self):
     def __get__(self, instance, owner_cls):
         if instance is None:
             # This is an unbound descriptor on the class
-            self._parent_cls = owner_cls
+            # NOTE! _encl_cls is not threadsafe
```

Comment:
It's actually worse. ``_encl_cls`` isn't even subclass safe. Consider:

```python
from astropy.utils.data_info import DataInfo

class A:
    info = DataInfo()

class B(A):
    pass

binfo = B.info
print(binfo._encl_cls)

ainfo = A.info
print(ainfo._encl_cls)

print(binfo._encl_cls) 
```
```
<class '__main__.B'>
<class '__main__.A'>
<class '__main__.A'>
```

I'm keeping it in this PR because it's in the old code, but we should definitely refactor to make this safe!

Reply:
@mhvk, I think this is a bug. Unfortunately it's pretty deeply embedded. ``_parent_cls`` is used in many places throughout astropy!